packages/pwa-kit-dev/src/utils/upload2.ts

```diff
@@ -0,0 +1,257 @@
+
+
+interface Credentials {
+    username: string
+    api_key: string
+}
+
+
+interface CloudAPIClientOpts {
+    credentials: Credentials,
+    origin?: string
+    fetch?: Function
+}
+
+
+interface StringMap {[key: string]: string;}
+
+
+interface Bundle {
+    message: string
+    encoding: string
+    data: string
+    ssr_parameters: object
+    ssr_only: string[]
+    ssr_shared: string[]
+}
+
+export const getPkgJSON = async () => {
+    const data = await readFile(path.join(__dirname, '..', '..', 'package.json'))
+    return JSON.parse(data.toString('utf-8'))
+}
+
+
+export class CloudAPIClient {
+    private opts: Required<CloudAPIClientOpts>
+
+    constructor(params: CloudAPIClientOpts) {
+        this.opts = {
+            origin: params.origin || DEFAULT_CLOUD_ORIGIN,
+            fetch: params.fetch || _fetch,
+            credentials: params.credentials,
+        }
+    }
+
+    private getAuthHeader(): StringMap {
+        const {username, api_key} = this.opts.credentials
+        const encoded = Buffer.from(`${username}:${api_key}`, 'binary').toString('base64')
+        return {'Authorization': `Basic ${encoded}`}
+    }
+
+    private async getHeaders(extras: StringMap): Promise<StringMap> {
+        const pkg = await getPkgJSON()
+        return {
+            'User-Agent': `progressive-web-sdk#${pkg.version}`,
+            ...this.getAuthHeader(),
+            ...extras,
+        }
+    }
+
+    private async throwForStatus(res) {
+        if (res.status < 400) {
+            return
+        }
+
+        let error
+        try {
+            error = await res.json()
+        } catch (err) {
+            error = {}  // Cloud doesn't always return JSON
+        }
+
+        throw new Error(
+            [
+                `HTTP ${res.status}`,
+                error.message || res.body,
+                `For more information visit ${error.docs_url || DEFAULT_DOCS_URL}`
+            ].join('\n')
+        )
+    }
+
+    async push(bundle: Bundle, projectSlug: string, target?: string) {
+        const base = `api/projects/${projectSlug}/builds/`
+        const pathname = target ? base + `${target}/` : base
+        const url = new URL(this.opts.origin)
+        url.pathname = pathname
+
+        const body = Buffer.from(JSON.stringify(bundle))
+        const headers = await this.getHeaders({'Content-Length': body.length.toString()})
+
+        const res = await this.opts.fetch(url.toString(), {
+            body,
+            method: 'POST',
+            headers,
+        })
+        await this.throwForStatus(res)
+        return await res.json()
+    }
+}
+
+export const defaultMessage = (gitInstance: git = git): string => {
+    try {
+        return `${gitInstance.branch()}: ${gitInstance.short()}`
+    } catch (err) {
+        if (err.code === 'ENOENT') {
+            console.log('Please run "git init" to initialize a new Git repository.')
```

Comment:
Would you consider a log statement here? Something like:

```suggestion
            console.log('Using default bundle message as no message was provided and not in a Git repo.')
```